src/program.cpp

```diff
@@ -483,6 +483,38 @@ void spir_binary::insert_descriptor_map(const spir_binary &other)
     m_dmaps_text += other.m_dmaps_text;
 }
 
+bool spir_binary::get_capabilities(std::vector<spv::Capability> &capabilities) const {
+    // Callback for receiving parsed instructions.
+    // The `user_data` parameter will be a pointer to the vector of
+    // capabilities (we cannot use a lambda capture for this as it prevents the
+    // lambda from being able to be converted to a function pointer).
+    auto parse_inst = [](void *user_data,
+                         const spv_parsed_instruction_t *inst) {
+        // Stop parsing at first instruction that is not an OpCapability.
+        if (inst->opcode != spv::Op::OpCapability) {
+            return SPV_END_OF_STREAM;
+        }
+
+        // Add the capability to the list.
+        uint32_t capability = inst->words[inst->operands[0].offset];
+        auto capabilities =
+            reinterpret_cast<std::vector<spv::Capability>*>(user_data);
+        capabilities->push_back(static_cast<spv::Capability>(capability));
+
+        return SPV_SUCCESS;
+    };
+
+    // Parse the SPIR-V binary to build the list of required capabilities.
+    spv_result_t result =
+        spvBinaryParse(m_context, &capabilities, m_code.data(), m_code.size(),
```

Comment:
I'd like to be able to cut the dependency on SPIR-V tools as a build option in the future which is why I haven't landed yet one of my patches that uses SPIR-V parsing to implement `CL_KERNEL_COMPILE_WORK_GROUP_SIZE`. Fine for now but good to keep in mind so we don't make this (too much) harder to do.

Reply:
Since `OpCapability` is always the first instruction(s) in a module it'd probably be easy enough to manually parse the binary for this information. I can do this in a follow-up patch now if you like?

Reply:
Nah, don't worry about that. I think the approach I would go for would probably be to completely disable the online capability checks in that case and to provide a way to encode the requirements as part of the build flow. The use-case I have in mind here is to be able to embed clvk into existing applications with minimal footprint.